ddtrace/appsec/internal/writer.py

```diff
@@ -0,0 +1,286 @@
+
+import attr
+import tenacity
+
+from ddtrace.appsec.internal.events import Event
+from ddtrace.internal import _encoding
+from ddtrace.internal import _rand
+from ddtrace.internal import agent
+from ddtrace.internal import atexit
+from ddtrace.internal import compat
+from ddtrace.internal import forksafe
+from ddtrace.internal import periodic
+from ddtrace.internal import service
+from ddtrace.utils.time import StopWatch
+
+
+if TYPE_CHECKING:
+    from ddtrace.internal.dogstatsd import DogStatsd
+
+log = logging.getLogger(__name__)
+
+DEFAULT_URL = agent.get_trace_url() + "/appsec/proxy/"
+DEFAULT_TIMEOUT = 2.0  # seconds
+DEFAULT_RETRY_ATTEMPS = 3
+DEFAULT_MAX_PAYLOAD_SIZE = 4 * 1024 * 1024  # b
+DEFAULT_FLUSH_INTERVAL = 3.0  # seconds
+
+
+class BaseEventWriter(object):
+    def write(self, events):
+        # type: (Iterable[Event]) -> None
+        """Queue AppSec events to the writer."""
+        raise NotImplementedError
+
+    def flush(self, timeout=None):
+        # type: (Optional[float]) -> None
+        """Send events and wait for completion."""
+        raise NotImplementedError
+
+
+class NullEventWriter(BaseEventWriter):
+    def write(self, events):
+        # type: (Iterable[Event]) -> None
+        pass
+
+    def flush(self, timeout=None):
+        # type: (Optional[float]) -> None
+        pass
+
+
+class HTTPEventEncoderV1(_encoding.ListBufferedEncoder):
+
+    content_type = "application/json"
+
+    def encode_item(self, item):
+        return json.dumps(attr.asdict(item), separators=(",", ":")).encode("ascii")
+
+    def encode(self):
+        data = self.get()
+        if data:
+            return (
+                b'{"protocol_version":1,"idempotency_key":"'
+                + str(_rand.rand64bits()).encode("ascii")
+                + b'","events": ['
+                + b",".join(data)
+                + b"]}"
+            )
+
+
+class HTTPRequestFailed(Exception):
+    """
+    Exception raised when the events payload cannot be sent.
+    """
+
+
+class HTTPEventWriter(periodic.PeriodicService, BaseEventWriter):
+    """
+    Batch events to an HTTP endpoint using AppSec intake protocol.
+    """
+
+    def __init__(
+        self,
+        url=DEFAULT_URL,  # type: str
+        api_key=None,  # type: Optional[str]
+        timeout=DEFAULT_TIMEOUT,  # type: float
+        retry_attemps=DEFAULT_RETRY_ATTEMPS,  # type: int
+        max_payload_size=DEFAULT_MAX_PAYLOAD_SIZE,  # type: int
+        flush_interval=DEFAULT_FLUSH_INTERVAL,  # type: float
+        dogstatsd=None,  # type: Optional[DogStatsd]
+    ):
+        # type: (...) -> None
+        super(HTTPEventWriter, self).__init__(interval=flush_interval)
+        self._retry_upload = tenacity.Retrying(
+            wait=tenacity.wait_random_exponential(multiplier=0.5),
+            stop=tenacity.stop_after_attempt(retry_attemps),
+            retry=tenacity.retry_if_exception_type((compat.httplib.HTTPException, OSError, IOError)),
+        )
```

Comment:
(aside) Given that we have similar logic in other places, I wonder if it would make sense to have e.g. a `Retriable` MixIn to throw into the mix